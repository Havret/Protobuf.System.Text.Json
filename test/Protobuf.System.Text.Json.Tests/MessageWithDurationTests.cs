using System.Text.Json;
using System.Text.Json.Protobuf.Tests;
using Google.Protobuf.WellKnownTypes;
using Protobuf.System.Text.Json.Tests.Utils;
using SmartAnalyzers.ApprovalTestsExtensions;
using Xunit;

namespace Protobuf.System.Text.Json.Tests;

public class MessageWithDurationTests
{
    [Fact]
    public void Should_serialize_message_with_duration()
    {
        // Arrange
        var msg = new MessageWithDuration
        {
            DurationProperty = new Duration {Seconds = 256, Nanos = 512}
        };

        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }

    [Fact]
    public void Should_serialize_message_with_duration_when_value_not_set()
    {
        // Arrange
        var msg = new MessageWithDuration();

        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }

    [Fact]
    public void Should_serialize_and_deserialize_message_with_duration()
    {
        // Arrange
        var msg = new MessageWithDuration
        {
            DurationProperty = new Duration {Seconds = 256, Nanos = 500}
        };
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithDuration>(serialized, jsonSerializerOptions);

        // Assert
        Assert.NotNull(deserialized);
        Assert.Equal(deserialized, msg);
    }

    [Fact]
    public void Should_serialize_and_deserialize_message_with_duration_when_value_is_not_set()
    {
        // Arrange
        var msg = new MessageWithDuration();
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithDuration>(serialized, jsonSerializerOptions);

        // Assert
        Assert.NotNull(deserialized);
        Assert.Equal(deserialized, msg);
    }

    [Fact]
    public void Should_serialize_duration_property_as_complex_object_when_TreatDurationAsTimeSpan_option_set_to_false()
    {
        // Arrange
        var msg = new MessageWithDuration
        {
            DurationProperty = new Duration {Seconds = 256, Nanos = 512}
        };

        var jsonSerializerOptions = new JsonSerializerOptions();
        jsonSerializerOptions.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;
        jsonSerializerOptions.AddProtobufSupport(options =>
        {
            options.TreatDurationAsTimeSpan = false;
        });

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }
    
    [Fact]
    public void Should_deserialize_duration_property_from_complex_object_when_TreatDurationAsTimeSpan_option_set_to_false()
    {
        // Arrange
        var msg = new MessageWithDuration
        {
            DurationProperty = new Duration {Seconds = 256, Nanos = 512}
        };

        var jsonSerializerOptions = new JsonSerializerOptions();
        jsonSerializerOptions.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;
        jsonSerializerOptions.AddProtobufSupport(options =>
        {
            options.TreatDurationAsTimeSpan = false;
        });

        // Act
        var payload = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithDuration>(payload, jsonSerializerOptions);

        // Assert
        Assert.NotNull(deserialized);
        Assert.Equal(deserialized, msg);
    }
}