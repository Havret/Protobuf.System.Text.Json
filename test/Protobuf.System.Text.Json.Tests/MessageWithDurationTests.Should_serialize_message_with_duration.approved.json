{
  "durationProperty": "00:04:16.0000005"
}